samza-core/src/main/java/org/apache/samza/system/inmemory/InMemorySystemFactory.java

```diff
@@ -31,20 +35,28 @@
  * Initial draft of in-memory {@link SystemFactory}. It is test only and not meant for production use right now.
  */
 public class InMemorySystemFactory implements SystemFactory {
-  private static final InMemoryManager MEMORY_MANAGER = new InMemoryManager();
+  private static Map<Integer, InMemoryManager> jobToManager = new HashMap<>();
 
   @Override
   public SystemConsumer getConsumer(String systemName, Config config, MetricsRegistry registry) {
-    return new InMemorySystemConsumer(MEMORY_MANAGER);
+    jobToManager.putIfAbsent(parseJobId(config), new InMemoryManager());
+    return new InMemorySystemConsumer(jobToManager.get(parseJobId(config)));
   }
 
   @Override
   public SystemProducer getProducer(String systemName, Config config, MetricsRegistry registry) {
-    return new InMemorySystemProducer(systemName, MEMORY_MANAGER);
+    jobToManager.putIfAbsent(parseJobId(config), new InMemoryManager());
+    return new InMemorySystemProducer(systemName, jobToManager.get(parseJobId(config)));
   }
 
   @Override
   public SystemAdmin getAdmin(String systemName, Config config) {
-    return new InMemorySystemAdmin(MEMORY_MANAGER);
+    jobToManager.putIfAbsent(parseJobId(config), new InMemoryManager());
+    return new InMemorySystemAdmin(jobToManager.get(parseJobId(config)));
+  }
+
+  private Integer parseJobId(Config config) {
+    Preconditions.checkNotNull(config.get(JobConfig.JOB_ID()));
```

Comment:
From Above:

Replaced jobId with test.id. The guarantees of this param is, if configured it will provide an independent instance of InMemorySystem to run, if not the application uses a default static InMemorySystem which is shared across all the classes

Additionally:
- Test Framework will generate an random TestId for an application, we don't expose this to users